pkg/storage/sst_writer.go

```diff
@@ -316,3 +318,57 @@ func (*MemFile) Sync() error {
 func (f *MemFile) Data() []byte {
 	return f.Bytes()
 }
+
+// AccountedMemFile wraps a MemFile so that can account for the memory it uses
+// in the given monitor if one is provided.
+type AccountedMemFile struct {
+	MemFile
+
+	ctx    context.Context
+	memAcc *mon.BoundAccount
+}
+
+// MakeAccountedMemFile creates an AccountedMemFile that can keep track of its
+// allocations towards the memory monitor provided.
+// Memory written to this buffer will be freed upon Close().
+func MakeAccountedMemFile(ctx context.Context, memMon *mon.BytesMonitor) (f AccountedMemFile) {
+	f.ctx = ctx
+	if memMon != nil {
+		memAcc := memMon.MakeBoundAccount()
+		f.memAcc = &memAcc
+	}
+
+	return
+}
+
+// Write implements the writeCloseSyncer interface. It accounts for memory it
+// uses.
+func (f *AccountedMemFile) Write(p []byte) (n int, err error) {
+	oldCap := f.MemFile.Cap()
+	n, err = f.MemFile.Write(p)
+	if err != nil {
+		return n, err
+	}
+
+	// Although this accounts for the memory after allocating it, that should
+	// generally be fine since we'll catch it. If this is not fine, we likely have
+	// larger issues and are very close to OOMing.
+	if f.memAcc != nil {
+		newCap := f.MemFile.Cap()
+		log.Infof(f.ctx, "growing by %d", int64(newCap-oldCap))
+		if err := f.memAcc.Grow(f.ctx, int64(newCap-oldCap)); err != nil {
+			return n, err
```

Comment:
crashing is obviously worse than erring here, sure, but ideally I'd say we could do even better, and make flushes budget-aware instead?

Total strawman that I haven't thought through: have `Write` flip a cheap-to-check bool. Move this grow attempt to the adding loop, inside a check of that bool, and trigger a flush if it fails. 

We could even avoid the check-after-growing-cap case that way, I think, by determining if the _next_ write could push it over e.g. if blocksize > cap-len, then we can suspect next write will grow cap, so we could go ahead and ask for more now and flush if we get denied. If we pad that by some trailer size, then I think we don't need to worry about flushing the trailer causing a realloc to grow cap, though I don't know if we can predict that perfectly since it'll depend on index size. Hmm.

Reply:
I took a look at that approach briefly and found that it coupled the batcher pretty tightly with `sstable.Writer` specifics pretty quickly which I originally wanted to avoid. One option here could be to have a pessimistic estimate (e.g. 1.5 x block size). AddMVCCKey can grow by this estimate if it detects that the next write will grow the buffer, and if that fails flush. If the Grow succeeds, we can shrink the buffer by our estimate and grow by the actual size the buffer increased on Write. That way the memory accounting stays up to date most of the time.